### PR TITLE
[WIP] Restructuring Testing Chapter 

### LINKED ISSUE #1103: Restructuring the chapter "Testing"
### Summary

The chapter [Testing](https://the-turing-way.netlify.app/testing/testing.html) needs to be restructured, it has to be sliced into sub-sections and needs some touch-up.

### What needs to be done?

Few proposed ideas (as of now)-

- At first glance, the chapter looks tremendously huge. Generally, short articles attract more audience and are easy to read. So, I will divide the chapter into sub-sections first. 

- Changing the python code written in between to a more-readable python-format.

### Updates

Work in Progress.

### PR DESCRIPTION
**Pull request by Yash-Varshney**
### Summary

Restructured the whole chapter `Testing` into different sub-chapters. Also, corrected grammatical mistakes and made it more formal.
Fixes #1103 

### List of changes proposed in this PR (pull-request)

- Restructured the chapter into sub-sections.

-  Made the python code more readable.

- Removed grammatical errors and made it more formal.

### What should a reviewer concentrate their feedback on?

- [ ] Format of the chapter.

- [ ] More suggestions to restructure.

- [ ]  Adding more information like writing `Benefits of Smoke test` and `tips for writing Smoke test`?

### Acknowledging contributors

- [x] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file.
- [ ] The following people should be added to the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file:

### CODE REVIEW

**Comment by Yash-Varshney** [comment]
@malvikasharan please lookout, if the format looks good and if you have any suggestions.

**Comment by malvikasharan** [comment]
Hi @Yash-Varshney, I have glanced through this PR. In general it looks good, but I wonder if we can still break the longer sections down to more modular subsections. Also, what do you think about the table of content? Does it seem alright to be on the intro page or should that move with the content?

**Comment by Yash-Varshney** [comment]
@malvikasharan regarding the `table of content` I thought it will be best to have the complete overview of the whole chapter in the first page itself. So that, if the reader wants to know a particular topic, he/she can directly point at it, also for every sub-section, there seems no need of having a table of content because it is automatically generated at the top-right corner itself.

**Comment by Yash-Varshney** [comment]
Also, I was thinking of breaking the `types of testing` section into two. It seems quite big.
Thanks for reviewing :)

**Comment by malvikasharan** [comment]
> @malvikasharan regarding the `table of content` I thought it will be best to have the complete overview of the whole chapter in the first page itself. So that, if the reader wants to know a particular topic, he/she can directly point at it

If the subchapters are small enough that can be read without scrolling too much, they definitely don't need TOC. 
I am also not in favor of a TOC in the landing page. I suppose the TOC itself can be shortened, of a summary of what to expect in each subchapter. This is because in the Netlify build we have a quick link to each chapter and subsection. You can leave that for now, and I can help update that after the subchapters are modularised.

**Comment by Yash-Varshney** [comment]
Sure @makvikasharan. I also thought the same way about a very long TOC and thought of introducing collapsible switch for every subchapter on the landing front page. But, I was not able to provide link for clicking directly onto it. 
It would be awesome to see that new features of netifly. Looking forward to it.
Thanks

**Comment by Yash-Varshney** [comment]
> Sure @malvikasharan. I also thought the same way about a very long TOC and thought of introducing collapsible switch for every subchapter on the landing front page. But, I was not able to provide link for clicking directly onto it. 
> It would be awesome to see that new features of netifly. Looking forward to it.
> Thanks

**Comment by malvikasharan** [comment]
Hi @Yash-Varshney, we are integrating standard format for each chapter as discussed in #1174, in case you have the time and interest, please feel free to edit these recommendations into the testing chapter. Also, please update your local branch as we have pushed a massive update in the main repo.

**Comment by Yash-Varshney** [comment]
Sure @malvikasharan. I am more than happy in doing that :)

File: book/content/testing/02(b)/description.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+<a name="Unit_testing_tips"></a>
+### Unit testing tips
+
+- Many testing frameworks have tools specifically geared towards writing and running unit tests.
+- Isolate the development environment from the test environment.
+- Write test cases that are independent of each other. For example, if a unit A utilises the result of another unit B supplies you should test unit A with a [test double](#Use_test_doubles_stubs_mocking_where_appropriate), rather than actually calling the unit B. If you don't do this your test failing may be due to a fault in either unit A *or* unit B, making the bug harder to trace.
+- Aim at covering all paths through a unit. Pay particular attention to loop conditions.
+- In addition to writing cases to verify the behaviour, write cases to ensure the performance of the code. For example, if a function that is supposed to add two numbers takes several minutes to run there is likely a problem.
```
Comment (bouweandela) [inline book/content/testing/02(b)/description.md:50]:
```suggestion
- In addition to writing cases to verify the behaviour, write cases to ensure the performance of the code. For example, if a function that is supposed to add two numbers takes several minutes to run, there is likely a problem.
```

File: book/content/testing/02(c)/description_cont.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
+<a name="Regression_testing"></a>
+## Regression testing
+
+Regression testing is a style of testing that focuses on retesting after changes are made. The results of tests after the changes are compared to the results before, and errors are raised if these are different. Regression testing is intended to ensure that changes (enhancements or defect fixes) to the software have not adversely affected it. The likelihood of any code change impacting functionalities that are not directly associated with the code is always there and it is essential that regression testing is conducted to make sure that fixing one thing has not broken another. Regression testing can be performed during any level of testing (unit, integration, system, or acceptance) but it is most relevant during system testing. Any test can be reused, and so any test can become a regression test.
+
+Regression testing is obviously especially important in team working, but it is surprisingly easy to break your own code without noticing it, even if you are working on your own. And because regression testing is next to impossible to do satisfactorily by hand (it's simply too tedious), it's an obvious case for automation.
```
Comment (bouweandela) [inline book/content/testing/02(c)/description_cont.md:6]:
```suggestion
Regression testing is obviously especially important when working in a team, but it is surprisingly easy to break your own code without noticing it, even if you are working on your own. And because regression testing is next to impossible to do satisfactorily by hand (it's simply too tedious), it's an obvious case for automation.
```

File: book/content/testing/02(c)/description_cont.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  if (array1.size() != array2.size()):
+    error( 'The arrays have different sizes!' )
+
+  output = array1 + array2
+
+  if (output.size() != array1.size()):
+    error( 'The output array has the wrong size!'' )
+
+  return output
+```
```
Comment (bouweandela) [inline book/content/testing/02(c)/description_cont.md:49]: This is not Python code, but some other language

File: book/content/testing/02(c)/description_cont.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+Advantages of runtime testing:
+
+- Run within the program, so can catch problems caused by logic errors or edge cases.
+- Makes it easier to find the cause of the bug by catching problems early.
+- Catching problems early also helps prevent them from escalating into catastrophic failures. It minimises the blast radius.
```
Comment (bouweandela) [inline book/content/testing/02(c)/description_cont.md:55]:
```suggestion
- Catching problems early also helps prevent them from escalating into catastrophic failures.
```

File: book/content/testing/04/testing-checklist.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+<a name="Checklist"></a>
+## Checklist
+
+This checklist contains a lot of items. As [mentioned](#Write_tests_any_tests) it is far better to do some of the items than none of them. Do not be discouraged if this list of tasks seems insurmountable.
```
Comment (bouweandela) [inline book/content/testing/04/testing-checklist.md:4]:
```suggestion
This checklist contains a lot of items. As [mentioned](#Write_tests_any_tests) before, it is far better to do some of the items than none of them. Do not be discouraged if this list of tasks seems insurmountable.
```

File: book/content/testing/04/testing-checklist.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+- [ ] Add runtime tests into your project.
+
+<a name="Good_practice_checks"></a>
+### Good practice checks
+
+- [ ] Document the tests and how to run them.
+  - [ ] Write scripts to set up and configure any resources that are needed to run the tests.
+- [ ] Pick and make use of a testing framework.
+- [ ] Run the tests regularly.
+  - [ ] Automate the process of running tests. Consider making use of continuous integration (see continuous integration chapter) to do this.
```
Comment (bouweandela) [inline book/content/testing/04/testing-checklist.md:24]: > ... (see continuous integration chapter) ...

Add a link?

File: book/content/testing/04/testing-checklist.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+- [ ] Check the code coverage of your tests and try to improve it.
+- [ ] Engage in code review with a partner.
```
Comment (bouweandela) [inline book/content/testing/04/testing-checklist.md:26]: > - [ ] Engage in code review with a partner.

Add a link to the code review chapter? Is this related to testing?

File: book/content/testing/05/testing-resources.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+- [Talk by Chrys Woods](https://drive.google.com/file/d/1CBTAhCVixccui1DjeUT13qh6ga5SDXjl/view) [**Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License**](https://chryswoods.com/main/copyright.html)
+
+### Materials used: Test driven development
+
+- [Testing your software](https://software.ac.uk/resources/guides/testing-your-software) **Creative Commons Attribution-NonCommercial 3.0 License.**
+- [Why bother](http://soundsoftware.ac.uk/unit-testing-why-bother/) **Creative Commons Attribution-NonCommercial 3.0 License.**
+
+### Materials used: glossary
+
+- [Netherlands eScience centre](https://guide.esciencecenter.nl/best_practices/testing.html) **Creative Commons Attribution 4.0 International License**
```
Comment (bouweandela) [inline book/content/testing/05/testing-resources.md:110]:
```suggestion
- [Netherlands eScience centre](https://guide.esciencecenter.nl/#/best_practices/testing) **Creative Commons Attribution 4.0 International License**
```

File: book/content/testing/05/testing-resources.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
+<a name="Resources"></a>
+## Resources
+
+
+<a name="What_to_learn_next"></a>
+### What to learn next
+
+Try reading the chapter on reproducible computational environments and then the chapter on continuous integration. The chapter on reviewing outlines how you can further strengthen your code base by adding a formal reviewing stage to your development workflow.
```
Comment (bouweandela) [inline book/content/testing/05/testing-resources.md:8]: Maybe add links to the respective chapters here?

**Review by bouweandela — CHANGES_REQUESTED** [review]
@malvikasharan @Yash-Varshney My apologies that it took me so long before I could have a look at this, it's rather a lot of changes to review in one pull request. I hope you find my feedback useful.
